server.go

```diff
@@ -102,18 +105,71 @@ var errServerClosed = errors.New("server was closed")
 // use in this server is greater than the provided limit, errPoolLimit is
 // returned.
 func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (socket *mongoSocket, abended bool, err error) {
+	return server.acquireSocketInternal(poolLimit, timeout, false, 0*time.Millisecond)
+}
+
+// AcquireSocketWithBlocking wraps AcquireSocket, but if a socket is not available, it will _not_
+// return errPoolLimit. Instead, it will block waiting for a socket to become available. If poolTimeout
+// should elapse before a socket is available, it will return errPoolTimeout.
+func (server *mongoServer) AcquireSocketWithBlocking(
+	poolLimit int, socketTimeout time.Duration, poolTimeout time.Duration,
+) (socket *mongoSocket, abended bool, err error) {
+	return server.acquireSocketInternal(poolLimit, socketTimeout, true, poolTimeout)
+}
+
+func (server *mongoServer) acquireSocketInternal(
+	poolLimit int, timeout time.Duration, shouldBlock bool, poolTimeout time.Duration,
+) (socket *mongoSocket, abended bool, err error) {
 	for {
 		server.Lock()
 		abended = server.abended
 		if server.closed {
 			server.Unlock()
 			return nil, abended, errServerClosed
 		}
-		n := len(server.unusedSockets)
-		if poolLimit > 0 && len(server.liveSockets)-n >= poolLimit {
-			server.Unlock()
-			return nil, false, errPoolLimit
+		if poolLimit > 0 {
+			if shouldBlock {
+				// Beautiful. Golang conditions don't have a WaitWithTimeout, so I've implemented the timeout
+				// with a wait + broadcast. The broadcast will cause the loop here to re-check the timeout,
+				// and fail if it is blown.
+				// Yes, this is a spurious wakeup, but we can't do a directed signal without having one condition
+				// variable per waiter, which would involve loop traversal in the RecycleSocket
+				// method.
+				// We also can't use the approach of turning a condition variable into a channel outlined in
+				// https://github.com/golang/go/issues/16620, since the lock needs to be held in _this_ goroutine.
+				waitDone := make(chan struct{})
+				timeoutHit := false
+				if poolTimeout > 0 {
+					go func() {
+						select {
+						case <-waitDone:
+						case <-time.After(poolTimeout):
+							// timeoutHit is part of the wait condition, so needs to be changed under mutex.
+							server.Lock()
```

Comment:
Also, I might be missing something here, but, 
the server lock is locked in line 124. 

The waiter on line 158 wait for a signal, but the signal is sent only if a socket is returned to the pool. 
The broadcast in line 151 won't be executed before an already used socket is released, as that code path tries to acquire lock that is is held by the parent goroutine (line 124) and releases on Wait() (line 158), no?

This is only after a cursory reading of the code. Please correct me if I am wrong.

Reply:
> the server lock is locked in line 124.

This lock on 148 is in a separate goroutine - there is a new goroutine spawned here to wait for the timeout to expire and signal the waiters if so.

> The broadcast in line 151 won't be executed before an already used socket is released, as that code path tries to acquire lock that is is held by the parent goroutine

Perhaps this is the most confusing part - the invocation of `Wait` on 158 actually releases the server mutex - the CondVar's `L` member is initialised to the server mutex on line 90, and the `Wait` method is defined to unlock `L`, wait for a signal, and then lock `L` again before returning control to the caller. 

So, in the case where a socket is released in time -
* The call to `AcquireSocket` locks the server mutex
* The call to `AcquireSocket` observes that there are no free connections
* The call to `AcquireSocket` unlocks the server mutex and waits for a signal (I believe this is atomic)
* The call to `RecycleSocket` locks the server mutex
* The call to `RecycleSocket` returns a connection to the free list
* The call to `RecycleSocket` broadcasts
* The call to `AcquireSocket` is woken from its `Wait`, but is still blocked waiting for the server mutex, so `Wait` does not yet return
* The call to `RecycleSocket` unlocks the server mutex
* The call to `AcquireSocket` locks the server mutex and returns control from `Wait`
* The call to `AcquireSocket` observes that there are now free connections and continues as usual.

In the case where a socket is not released on time -
* The call to `AcquireSocket` locks the server mutex
* The call to `AcquireSocket` observes that there are no free connections
* The call to `AcquireSocket` unlocks the server mutex and waits for a signal
* The timer goroutine hits the `time.After` timeout
* The timer goroutine takes the server mutex
* The timer goroutine sets the timeout flag
* The timer goroutine broadcasts
* The call to `AcquireSocket` is woken from its `Wait`, but is still blocked waiting for the server mutex, so `Wait` does not yet return
* The timer goroutine unlocks the server mutex
* The call to `AcquireSocket` locks the server mutex and returns control from `Wait`
* The call to `AcquireSocket` notices that the timeout flag has been set
* The call to `AcquireSocket` unlocks the mutex and errors out

I mean, this is concurrent code with mutexs and semaphores, so I'm practically guaranteed to be wrong, but this is my understanding of how this works and hopefully this answers your question.

Reply:
Right. I forgot that the Wait() releases the lock before yielding the goroutine.